documentation/specs/src/base-ledger/block-space-allocator.md

```diff
@@ -0,0 +1,185 @@
+in an abstract container dubbed the `TxBin`. A transaction $tx : tx \in M$ may 
+be **dumped** to a `TxBin`, resulting in a successful operation, or an error, 
+if $tx$ is **rejected** due to lack of space in the `TxBin` or if $tx$'s size 
+**overflows** (i.e. does not fit in) the `TxBin`. Block proposers continue 
+dumping transactions from $M$ into a `TxBin` $B$ until a rejection error is 
+encountered, or until there are no more transactions of the same type as $B$'s 
+in $M$. The `BlockSpaceAllocator` contains three `TxBin` instances, responsible 
+for holding decrypted, protocol and encrypted transactions.
+
+<img
+  src="images/block-space-allocator-bins.svg"
+  alt="block space allocator tx bins"
+  height="400"
+  width="500"
+  style="display: block; margin: 0 auto" />
+
+During occasional Namada protocol events, such as DKG parameter negotiation, 
+all available block space should be reserved to protocol transactions, 
+therefore the `BlockSpaceAllocator` was designed as a state machine, whose 
+state transitions depend on the state of Namada. The states of the 
+`BlockSpaceAllocator` are the following:
+
+1. `BuildingDecryptedTxBatch` - As the name implies, during this state the 
+decrypted transactions `TxBin` is filled with transactions of the same type. 
+Honest block proposers will only include decrypted transactions in a block at a 
+fixed height $H_0$ if encrypted transactions were available at $H_0 - 1$. The 
+decrypted transactions should be included in the same order of the encrypted 
+transactions of block $H_0 - 1$. Likewise, all decrypted transactions available 
+at $H_0$ must be included.
+2. `BuildingProtocolTxBatch` - In a similar manner, during this 
+`BlockSpaceAllocator` state, the protocol transactions `TxBin` is populated 
+with transactions of the same type. Contrary to the first state, allocation 
+stops as soon as the respective `TxBin` runs out of space for some 
+$tx_{Protocol} : tx_{Protocol} \in M$. The `TxBin` for protocol transactions is 
+allotted half of the remaining block space, after decrypted transactions have 
+been **allocated**.
+3. `BuildingEncryptedTxBatch` - This state behaves a lot like the previous 
+state, with one addition: it takes a parameter that guards the encrypted 
+transactions `TxBin`, which in effect splits the state into two sub-states. 
+When `WithEncryptedTxs` is active, we fill block space with encrypted 
+transactions (as the name implies); orthogonal to this mode of operation, there 
+is `WithoutEncryptedTxs`, which, as the name implies, does not allow encrypted 
+transactions to be included in a block. The `TxBin` for encrypted transactions 
+is allotted $\min(R,\frac{1}{3} MaxProposalBytes)$ bytes, where $R$ is the 
+block space remaining after allocating space for decrypted and protocol 
+transactions.
+4. `FillingRemainingSpace` - The final state of the `BlockSpaceAllocator`. Due 
+to the short-circuit behavior of a `TxBin`, on allocation errors, some space 
```

Comment:
we should probably state the reasoning behind filling up unused space with protocol txs only.